src/lib.rs

```diff
@@ -0,0 +1,350 @@
+    /// Which address version
+    pub version: Version,
+    /// Which public‑key type
+    pub pubkey_type: PubKeyType,
+    /// The raw public key hash bytes
+    pub pubkey_hash: Vec<u8>,
+}
+
+impl DecodedAddress {
+    /// The raw public key hash bytes
+    #[must_use]
+    pub fn pubkey_hash_bytes(&self) -> &[u8] {
+        &self.pubkey_hash
+    }
+
+    /// The hex representation of the public key hash
+    #[must_use]
+    pub fn pubkey_hash_hex(&self) -> String {
+        hex_encode(&self.pubkey_hash)
+    }
+}
+
+/// Errors that can occur during address decoding.
+#[derive(Error, Debug)]
+pub enum AddressDecodeError {
+    /// Invalid Bech32 structure or checksum
+    #[error("Bech32 error: {0}")]
+    Bech32(#[from] bech32::DecodeError),
+
+    /// HRP wasn’t `yp` or `rh`
+    #[error("unknown HRP: {0}")]
+    UnknownHrp(String),
+
+    /// Payload too short
+    #[error("payload too short (need at least version+public key type+hash algorithm)")]
+    TooShort,
+
+    /// First byte isn’t a known version code
+    #[error("unknown version code: 0x{0:02X}")]
+    UnknownVersion(u8),
+
+    /// Second byte isn’t a known public key type
+    #[error("unknown public key type code: 0x{0:02X}")]
+    UnknownPubKeyType(u8),
+
+    /// Digest length didn’t match the algorithm’s expectation
+    #[error("invalid digest length: got {got}, expected {expected}")]
+    InvalidHashLength { got: usize, expected: usize },
+}
+
+/// Decode a Bech32m address string back into its parts.
+///
+/// # Errors
+/// Returns `AddressDecodeError` on any failure:
+/// - Bad checksum or format
+/// - Unknown HRP
+/// - Invalid payload length
+/// - Unrecognized version, public key type or hash algorithm codes
+/// - Digest length mismatch
+pub fn decode_address(s: &str) -> Result<DecodedAddress, AddressDecodeError> {
+    // Decode Bech32
+    let (hrp, bytes) = decode(s)?;
+
+    // Map HRP to network enum
+    let network = Network::try_from(hrp.as_str())?;
+
+    // Validate minimum length for version + public key type + hash algorithm
+    if bytes.len() < 3 {
+        return Err(AddressDecodeError::TooShort);
+    }
+
+    // Parse version code
+    let v = bytes[0];
+    let version = Version::from_code(v).ok_or(AddressDecodeError::UnknownVersion(v))?;
+
+    // Parse public key type code
+    let p = bytes[1];
+    let pubkey_type = PubKeyType::from_code(p).ok_or(AddressDecodeError::UnknownPubKeyType(p))?;
+
+    // Extract and verify digest length
+    let pubkey_hash = bytes[2..].to_vec();
+    let expected = Hasher::DIGEST_LENGTH;
+    if pubkey_hash.len() != expected {
+        return Err(AddressDecodeError::InvalidHashLength {
+            got: pubkey_hash.len(),
+            expected,
+        });
+    }
+
+    Ok(DecodedAddress {
+        network,
+        version,
+        pubkey_type,
+        pubkey_hash,
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    /// Round‑trip test on Mainnet with SHA‑256 + `ML‑DSA 44`
```

Comment:
Lint: We can remove SHA256 from these comments because we're always using SHA256